### PR TITLE
Issues #22, #21, #20 multiple buildings, floors and chaining directions

### PR DESCRIPTION
**Pull request by NicholasNagy**
#20, #21, #22 Development:  

This pull request includes the following features: 

- A number picker for choosing the floor on the indoor maps
- Multiple floor maps for one indoor location
- Change room numbers for the appropriate floor
- Multiple buildings' indoor maps are now supported

- Now when a user wants to navigate from indoor to outdoor or vice versa, we do the navigation steps
- We also offer the ability to navigate back and forth between navigation steps

### CODE REVIEW

File: app/src/androidTest/java/com/droidhats/campuscompass/views/FloorFragmentTest.kt [hunk 1]
```diff
@@ -0,0 +1,154 @@
+package com.droidhats.campuscompass.views
+
+import androidx.navigation.testing.TestNavHostController
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.*
```
Comment (makchamp) [inline app/src/androidTest/java/com/droidhats/campuscompass/views/FloorFragmentTest.kt:6]: Change * import to specify specific imports

File: app/src/main/java/com/droidhats/campuscompass/NavHandler/IndoorNavStep.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import com.droidhats.campuscompass.models.Location
+import com.droidhats.campuscompass.models.NavigationRoute
+import com.droidhats.campuscompass.repositories.NavigationRepository
+
+class IndoorNavStep(override val location: IndoorLocation) : NavHandler() {
+
+    override fun getNavigationRoute() {
+        if (next == null) {
+            val navRoute = NavigationRoute(
+                IndoorLocation(
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/NavHandler/IndoorNavStep.kt:13]: why don't you put this all on one line?

File: app/src/main/java/com/droidhats/campuscompass/models/NavigationRoute.kt [hunk 3]
```diff
@@ -2,14 +2,19 @@ package com.droidhats.campuscompass.models
 
 import com.google.android.gms.maps.model.LatLng
 
-class NavigationRoute(
-    var origin: Location?,
-    var destination: Location?,
+open class NavigationRoute(
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/models/NavigationRoute.kt:5]: why don't you have this on one line?

File: app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt [hunk 4]
```diff
@@ -40,36 +43,40 @@ class IndoorLocationRepository private constructor(private val indoorLocationDao
             indoorLocationDao.deleteAllIndoor()
         }
         if (config == "debug" || indoorLocationDao.getOne().value == null) {
-            map.forEachBuilding { building ->
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt:43]: we might need to check if this method is still being used

File: app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt [hunk 5]
```diff
@@ -40,36 +43,40 @@ class IndoorLocationRepository private constructor(private val indoorLocationDao
...
 
         for (floorMap in building.getIndoorInfo().second) {
-            val inputStream: InputStream = context.assets.open(floorMap)
+            val inputStream: InputStream = context.assets.open(floorMap.value)
             val file: String = inputStream.bufferedReader().use { it.readText() }
             val mapProcessor: ProcessMap = ProcessMap()
             mapProcessor.readSVGFromString(file)
             val classes = mapProcessor.getClasses()
 
-            // todo: Consider the case where floor number is more than 1 digit
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt:58]: We need to fix this todo before the merge

File: app/src/main/java/com/droidhats/campuscompass/repositories/NavigationRepository.kt [hunk 6]
```diff
@@ -43,6 +46,7 @@ class NavigationRepository(private val application: Application) {
     private var sgwShuttleTimes: LiveData<List<ShuttleBusSGWEntity>>
     private var navigationRoute = MutableLiveData<NavigationRoute>()
     var routeTimes = MutableLiveData<MutableMap<String, String>>()
+    var navhandler: NavHandler? = null
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/repositories/NavigationRepository.kt:50]: why are there 2 empty lines here?

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt [hunk 7]
```diff
@@ -1,59 +1,284 @@
...
+            mapToDisplay = floormap
+        }
+
+        val inputStream: InputStream = requireContext().assets.open(mapToDisplay)
+        val mapProcessor: ProcessMap = ProcessMap()
+        var file: String = inputStream.bufferedReader().use { it.readText() }
+        file = mapProcessor.automateSVG(file, floorNum)
+
+        val classRoomToHighlight: String? = arguments?.getString("id")
+
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:233]: why the space here?

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt [hunk 8]
```diff
@@ -1,59 +1,284 @@
...
+            keys[index] = map.key
+            maps.add(map.value)
+            index++
+        }
+
+        val numberPicker: NumberPicker = root.findViewById(R.id.floorPicker)
+        numberPicker.minValue = 0
+        numberPicker.maxValue = keys.size - 1
+        numberPicker.wrapSelectorWheel = false
+
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:261]: why the space here?

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt [hunk 9]
```diff
@@ -68,23 +293,255 @@ class FloorFragment : Fragment() {
...
+            }
+            startAndEnd.first.lID == "" -> {
+                1 < startAndEnd.second.getFloorNumber()
+            }
+            else -> { // second.lID is an empty string
+                startAndEnd.first.getFloorNumber() < 1
+            }
+        }
+
+
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:337]: add comments here to explain the onclick listener methods

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt [hunk 10]
```diff
@@ -68,23 +293,255 @@ class FloorFragment : Fragment() {
...
+                        )
+                    }
+                }
+            }
+            intermediateTransportID = null
+            it.visibility = View.GONE
+            doneButton.visibility = View.VISIBLE
+            prevArrowButton.visibility = View.VISIBLE
+        }
+        prevArrowButton.setOnClickListener {
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:402]: space this out

File: app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt [hunk 11]
```diff
@@ -333,6 +417,130 @@ class ProcessMap {
...
+            }
+
+            // if a closest point was found do stuff
+            if(closestPoint != null) {
+                doForClosestPoint(closestPoint)
+            }
+        }
+    }
+
+    fun getNearestPathPoint(center: Pair<Double, Double>,
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt:507]: Add a comment here

File: app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt [hunk 12]
```diff
@@ -342,26 +550,100 @@ class ProcessMap {
...
+
+        val num: Int = pathPoints.size/8
+        val threads: MutableList<Thread> = mutableListOf()
+
+        for(point in pathPoints) {
+            vertices.add(Vertex(point, endPoint.getCenter()))
+        }
+
+        for (i in 0..7) {
+            println(i)
```
Comment (makchamp) [inline app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt:594]: remove this print line

**Comment by JeremiahTee** [comment]
> I left some comments in the code that need to be fixed before we can merge

Addressed these comments and implemented changes in 2a8cec2 during a multi-pair programming session over Teamviewer / Discord with @NicholasNagy @makchamp @KhaledBadran @zohalmir

File: app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt (hunk 12 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt:632]: Fixed

File: app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt (hunk 11 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/mapprocessor/ProcessMap.kt]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt (hunk 10 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:416]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt (hunk 9 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt]: Done

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt (hunk 8 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt (hunk 7 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:231]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/repositories/NavigationRepository.kt (hunk 6 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/repositories/NavigationRepository.kt]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt (hunk 5 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt:58]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/models/NavigationRoute.kt (hunk 3 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/models/NavigationRoute.kt]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/NavHandler/IndoorNavStep.kt (hunk 2 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/NavHandler/IndoorNavStep.kt]: Fixed

File: app/src/androidTest/java/com/droidhats/campuscompass/views/FloorFragmentTest.kt (hunk 1 above)
Comment (JeremiahTee) [reply app/src/androidTest/java/com/droidhats/campuscompass/views/FloorFragmentTest.kt]: Fixed

File: app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt (hunk 4 above)
Comment (NicholasNagy) [reply app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt:43]: true, It was not being used, so I removed it

File: app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt (hunk 5 above)
Comment (NicholasNagy) [reply app/src/main/java/com/droidhats/campuscompass/repositories/IndoorLocationRepository.kt:58]: It was actually already implemented, I just had forgotten to remove the todo

**Comment by NicholasNagy** [comment]
I addressed the rest of the comments that you had suggested @makchamp . 
I left some final dokka comments to the code, since the code was lacking.

**Comment by JeremiahTee** [comment]
I reviewed the changes made by @NicholasNagy and @makchamp and ran the test coverage. All is good. We can proceed to merge

**Pull request merged**